File: collatz.cpp
#include <iostream>
#include <algorithm>
using namespace std;
// keep hold of the data in a struct 
struct numbers {
     long long  num;
     long long steps;
} c[10];

// prototype
long long numSteps( long long n);

int  main()
{
     int size = 10; 
     long long max;
    cout << "Enter a positive number"<<endl;
    cin >> max;

    // fill array
    for( long long i = 0; i < size; i++)
    {
        c[i].steps = 0;
        c[i].num = 0;
    }
    // number with largest amount
    while(max != 0)
    {
         long long stepnum = numSteps(max); 
        bool dup = false;
         long long smallest = 0;
         long long small = c[0].steps;
         long long dupIndex = 0;
        // loop through the first size elements
        for( long long i = 0; i < size; i++)
        {
            if(c[i].steps < small)
            {
                small = c[i].steps;
                smallest = i;
            }
            // if one has duplicate, eliminate it
            if(c[i].steps == stepnum)
            {
                dup = true;
                dupIndex = i;
            }
        }
         long long index = smallest;
        if(stepnum > c[index].steps && dup == false)
        {
            c[index].steps = stepnum;
            c[index].num = max;
        }
        if(max < c[dupIndex].steps && dup == true)
        {
            c[dupIndex].num = max;
        }
        max--;
    }
    cout << "Sorted based on sequence length"<< endl;
    sort(c,c+size,[] (numbers a, numbers b){return a.steps > b.steps;});
    for(long long  i = 0; i< size; i++)
    {
        cout << c[i].num << "     "<< c[i].steps << endl;
    }
    cout << "Sorted based on integer length"<< endl;
    sort(c,c+size,[](numbers a, numbers b){return a.num > b.num;});
    for(long long  i =0; i < size;i++)
    {
        cout << c[i].num << "      " << c[i].steps << endl;
    }
}

long long numSteps(long long n)
{ 
    if(n == 1)
    {
        return 0; 
    }
    else if ( n & 1)
    {
        return numSteps((n<<1)+n+1) + 1; 
    }
    else
    { 
        return  numSteps(n>>1) + 1; 
    }  
}

File: collatz.pl
#!/usr/bin/perl
use diagnostics;
use warnings;
use feature 'say';
no warnings 'recursion';
my $max;
my $control;
my @num;
my @step;
my $steps;
my $dup;
my $dupIndex;
my $smallIndex;
my $small;
my $temp = 0;
my $size = 10; 
say "Please enter in a positive number";
$max = <STDIN>;
$control = $max;
#initialize array
for(my $i = 0; $i < $size;$i++)
{
    $step[$i] =0;
    $num[$i] = 0;
}
while($control > 1)
{
    $max = $control;
    $steps = collatz($max); 
    $max = $control;
    $dup = 0;
    $dupIndex = 0;
    $smallIndex = 0;
    $small = $step[0];
# update the smallest value
    for(my $i = 0; $i < $size; $i++)
    {
        if($step[$i] < $small)
        {
            $small = $step[$i];
            $smallIndex = $i;
        }
        if($step[$i] == $steps)
        {
            $dup = 1;
            $dupIndex = $i;
        }
    }
# if no duplicate is found add it to the list
    if($steps > $step[$smallIndex] and $dup == 0)
    {
        $step[$smallIndex] = $steps;
        $num[$smallIndex] = $max;
    }
# if a duplicate is found leave it out
    if($max < $num[$dupIndex] and $dup == 1)
    {
        $num[$dupIndex] = $max;
    }
    $control = $control - 1;
}
#bubble sort 
say "Sorted by sequence length";
for($j = 0; $j < $size-1; $j = $j + 1)
{
    for($k = 0; $k < $size-$j-1; $k = $k +1)
    {
        if($step[$k] > $step[$k +1])
        {
            $temp = $step[$k];
            $step[$k] = $step[$k+1];
            $step[$k+1] = $temp;

            $temp = $num[$k];
            $num[$k] = $num[$k+1];
            $num[$k+1] = $temp;
        }
    }
}
 
for(my $j = $size-1; $j >= 0; $j--)
{
    say $num[$j], "  ", $step[$j];
}
#bubble sort 
for($j = 0; $j < $size-1; $j = $j + 1)
{
    for($k = 0; $k < $size-$j-1; $k = $k +1)
    {
        if($num[$k] > $num[$k +1])
        {
            $temp = $num[$k];
            $num[$k] = $num[$k+1];
            $num[$k+1] = $temp;

            $temp = $step[$k];
            $step[$k] = $step[$k+1];
            $step[$k+1] = $temp;
        }
    }
}
say "sorted by integer length";
for(my $i = $size-1; $i >= 0; $i--)
{
    say $num[$i], "   ", $step[$i];
}

sub collatz{
    my ($n) = @_;
    if($n == 1) 
    {
        return 0; 
    }
    else
    {
        if($n & 1 == 1)
        {
            return 1 + collatz(($n<<1)+$n +1); 
        }
        else
        {
            return 1 + collatz($n>>1);
        }
    }
}
